ambari-server/src/main/java/org/apache/ambari/server/events/publishers/AgentCommandsPublisher.java

```diff
@@ -244,9 +244,12 @@ public static KerberosCommandParameterProcessor getInstance(String command, Clus
       KerberosServerAction.KerberosCommandParameters kerberosCommandParameters = new KerberosServerAction.KerberosCommandParameters(executionCommand);
 
       try {
-        Map<String, ? extends Collection<String>> serviceComponentFilter = getServiceComponentFilter(kerberosCommandParameters.getServiceComponentFilter());
+        Map<String, Collection<String>> serviceComponentFilter = getServiceComponentFilter(kerberosCommandParameters.getServiceComponentFilter());
 
         Set<ResolvedKerberosKeytab> keytabsToInject = kerberosKeytabController.getFilteredKeytabs(serviceComponentFilter, kerberosCommandParameters.getHostFilter(), kerberosCommandParameters.getIdentityFilter());
+        if (serviceComponentFilter != null) {
+          keytabsToInject.addAll(kerberosKeytabController.findKeytabsForServiceIdentities(executionCommand.getClusterName(), serviceComponentFilter));
```

Comment:
It seems like this will negate any host filter.  Have you tested regenerating keytab files for a specific host?  Also, does this break any operations related to cleaning up identities?

When I was testing this for a different patch, I believe that the `kerberosKeytabController.getFilteredKeytabs` call above found both service and user (headless) principals related to a given service.

Reply:
This is true - `kerberosKeytabController.getFilteredKeytabs` finds both service and user principals - only if the given service has any components on that host.
I can reproduce this issue with 2.7.1.0-163 (see description above)

This was not the case in the cluster where QE found the issue: they created the cluster with BP - topology validation was switched off - with a host group where `Druid Historical` was present but no `HDFS Client`

Reply:
See`org.apache.ambari.server.orm.dao.KerberosKeytabPrincipalDAO.findByFilter(KerberosKeytabPrincipalFilter filter)`: since the given filter is populated with `service name = HDFS` the underlying DB query will find only the principals for the `Datanode` (which does not need the HDFS user principal)

Reply:
Which means that we may convert everything in the service name filter to the principal filter and do not check the service name at all.